bootstrap/src/q.ml

```diff
@@ -0,0 +1,202 @@
+open Rudiments
+
+type 'a t = usize * 'a Stream.t * 'a list * 'a Stream.t
+
+let empty = (0, lazy Stream.Nil, [], lazy Stream.Nil)
+
+let exec (l, f, r, s) =
+  let rec rotate (l, f, r, s) = lazy begin
+    match f, r with
+    | _, [] -> assert false
+    | lazy Stream.Nil, y :: _ -> Lazy.force (Stream.push y s)
+    | lazy (Stream.Cons(elm, f')), y :: y' -> begin
+      let s' = Stream.push y s in
+      let f'' = rotate (l, f', y', s') in
+      Lazy.force (Stream.push elm f'')
+    end
+  end in
+  match s with
+  | lazy (Stream.Cons(_, s')) -> (l, f, r, s')
+  | lazy Stream.Nil -> begin
+    let f' = rotate (l, f, r, s) in
+    (l, f', [], f')
+  end
+
+let length (l, _, _, _) = l
+
+let is_empty (l, _, _, _) =
+  l = 0
+
+let hd (_, f, _, _) =
+  match f with
+  | lazy Stream.Nil -> halt "Empty q has no head"
+  | lazy (Stream.Cons(elm, _)) -> elm
+
+let tl (l, f, r, s) =
+  match f with
+  | lazy Stream.Nil -> halt "Empty q has no tail"
+  | lazy (Stream.Cons(_, f')) -> exec (pred l, f', r, s)
+
+let push_back (l, f, r, s) elm =
+  exec (succ l, f, elm :: r, s)
+
+let pop t =
+  hd t, tl t
+
+let pp pp_elm ppf (l, f, r, s) =
+  let open Format in
+  fprintf ppf "@[<h>";
+  fprintf ppf "(len=%a, f=%a, r=%a, s=%a)"
+  Usize.pp l (Stream.pp pp_elm) f (List.pp pp_elm) r (Stream.pp pp_elm) s;
+  fprintf ppf "@]"
+
+(*******************************************************************************
+ * Begin tests.
+ *)
+
+let%expect_test "empty" =
+  let open Format in
+  let ppt = (pp Usize.pp) in
+  printf "@[<h>";
+  let t = empty in
+  printf "empty = %a\n" ppt t;
+  printf "@]";
+
+  [%expect{| empty = (len=0, f=Nil, r=[], s=Nil) |}]
+
+let%expect_test "length" =
+  let open Format in
+  let ppt = (pp Usize.pp) in
+  printf "@[<h>";
+  let rec test_length_up_to i n t = begin
+    match i < n with
```

Comment:
I'd expect `*_up_to` to apply to an open bound (`match i <= n with`), in order to be consistent with the semantics of for loop bounds `to` and `downto`.  TBH I'm not crazy about the keyword names, but the open-open semantics are actually important, e.g. in the `downto` case, since otherwise there would be no way to express "down to 0 inclusive" on unsigned integers.

Reply:
Do you mean the naming like "test_length_up_to"? I'd be fine with just changing all those to fn.

Reply:
I'm fine with just `fn` here, or a change to the names (e.g. `test_length` here).

Reply:
Cool. Some of these comments apply to the stream module that I've already commited. I've fixed these in there as well and will make a final "Use consistent idioms" PR when we're done with deq.